### ISSUE TITLE
[gui] Hex view data error

### ISSUE THREAD

**Issue by Asteri5m** [skylot/jadx#2160]
### Issue details

When looking at the resource file, the hex view's data content is different from what I saw using the 010 Editor. When I use the algorithm in the program to decrypt, only the latter data is available
![image](https://github.com/skylot/jadx/assets/73411291/c5f41bc0-49f0-4b6a-8bda-fef74fba0e49)

### Jadx version

1.5.0

### Java version

21.0.3

### OS

- [X] Windows
- [ ] Linux
- [ ] macOS

**Comment by jpstotz** [comment]
Ist it possible to provide the APK file that contains the shown `libc.so`?  Or at least a screenshot of 7-Zip having the APK opened and showing the `libc.so` entry with size, compressed size and compression method.

**Comment by Asteri5m** [comment]
This is a virus, use with caution. Do not run on real machines.
pass:&nbsp;jadx_sample




asteri5m
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "skylot/jadx"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年4月23日(星期二) 下午3:36
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [skylot/jadx] [gui] Hex view data error (Issue #2160)





 
Ist it possible to provide the APK file that contains the shown libc.so?  Or at least a screenshot of 7-Zip having the APK opened and showing the libc.so entry with size, compressed size and compression method.

**Comment by jpstotz** [comment]
@Asteri5m Sorry I don't see any file. It seems like posting attachments via e-mail to Github issues is not possible.

**Comment by Asteri5m** [comment]
[41514E30ADB73AC904296DFBF4C029A1.zip](https://github.com/skylot/jadx/files/15088601/41514E30ADB73AC904296DFBF4C029A1.zip)

**Comment by jpstotz** [comment]
The main problem is that the HexArea that display the hex data is getting the data to be displayed from a String that is then decoded as UTF-8 string to a byte[]:

https://github.com/skylot/jadx/blob/b80f32a36f4a7df1562bb426a175ffdd8aba98c6/jadx-gui/src/main/java/jadx/gui/ui/codearea/HexArea.java#L51

For binary files, the unicode decoding can modify the data which is most likely the reason you have described in this issue. Binary data should never be saved in a String...

I think for a JResource node with ZipRef it would be more feasible to directly load the raw file data from the ZIP file instead of using the loaded data from the node.

I made a quick test using the following code. @skylot Do you see any side effects when loading the hex data directly from the ZIP file?
```
	public void load() {
		byte[] bytes = null;
		if (binaryNode instanceof JResource) {
			JResource jResource = ((JResource) binaryNode);
			ResourceFile.ZipRef zipRef = jResource.getResFile().getZipRef();
			if (zipRef != null) {
				try (ZipFile zipFile = new ZipFile(zipRef.getZipFile())) {
					ZipEntry entry = zipFile.getEntry(zipRef.getEntryName());
					if (entry == null) {
						throw new RuntimeException("Zip entry not found: " + zipRef);
					}
					if (ZipSecurity.isValidZipEntry(entry)) {
						try (InputStream inputStream = ZipSecurity.getInputStreamForEntry(zipFile, entry)) {
							bytes = inputStream.readAllBytes();
						}
					}
				} catch (IOException e) {
					throw new RuntimeException(e);
				}
			}
		}
		if (bytes == null) {
			bytes = binaryNode.getCodeInfo().getCodeStr().getBytes(StandardCharsets.UTF_8);
		}
....
```

**Comment by skylot** [comment]
> Binary data should never be saved in a String

I agree, this is should be avoided at all.

> Do you see any side effects when loading the hex data directly from the ZIP file?

I think this is fine, but it is better to not duplicate such big chunk of code and reuse loading code from https://github.com/skylot/jadx/blob/f5accc84643c40da1148cd753a8118019a381d16/jadx-core/src/main/java/jadx/api/ResourcesLoader.java#L56
Another approach is to make loading similar to `ImagePanel`. It is store byte array in `ResContainer` with type `DECODED_DATA` https://github.com/skylot/jadx/blob/6e8affcbdc917965379d524f245ba96d107d5be5/jadx-gui/src/main/java/jadx/gui/ui/panel/ImagePanel.java#L39
and loading of data is done here:
https://github.com/skylot/jadx/blob/f5accc84643c40da1148cd753a8118019a381d16/jadx-core/src/main/java/jadx/api/ResourcesLoader.java#L117-L118

Anyway, both ways are fine, because all resource code loading is a mess and should be simplified and refactored :rofl:

**Comment by Asteri5m** [comment]
It seems that the problem has been resolved. I hope to update this bug and release a new release as soon as possible. This feature is still very practical. Like

**Comment by jpstotz** [comment]
From my perspective the direct way using `ResourcesLoader.decodeStream` would be the best as the other ways seem to base on `ResourceLoader.loadContent` which also considers the type of the resource which is not what we want for a hex view.

**Comment by jpstotz** [comment]
You don't have to wait for a new version to be released. Once my PR is merged into the code base the Github CI system will build a new unstable version which you can then download here: [unstable download](https://nightly.link/skylot/jadx/workflows/build-artifacts/master) (also mentioned in the main project [README.md](https://github.com/skylot/jadx/blob/master/README.md#Download)).

**Comment by skylot** [comment]
> resource code loading is a mess and should be simplified and refactored 🤣

Looks like @AndreiKud already start refactoring in PR #2165  :eyes:
